# Coding Exercises (Part 5)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Intermediate

### Exercise 5: Sorting & Filtering 

In [2]:
# run the cell!
import pandas as pd

In [3]:
# run the cell!
listings = pd.read_csv("listings.csv")

In [4]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

39. __Sort__ the __listings__ DataFrame by the __"Market_Cap"__ column from __high to low__ and __save__ the new order! __Fill in the gaps__!

In [6]:
listings.sort_values(by = 'Market_Cap',ascending=False,inplace=True)


__Inspect__ listings! Which Company has the __highest Market Capitalization__ (Name)?

In [7]:
#run the cell!
res = listings.loc[listings['Market_Cap'].idxmax()]['Name']
res

'Apple Inc.'

The highest Market Capitalization has ... Apple!

40. __Sort__ listings by __IPO_Year__ (ascending) and __Market_Cap__ (descending) and save the new order. 

In [8]:
listings.sort_values(by =['IPO_Year','Market_Cap'],ascending=[True,False],inplace=True)


__Inspect__ listings! From all Companies that were listed in __1983__, which one has the __highest Market Capitalization__?

In [24]:
#run the cell!

companies_1983 = listings[listings['IPO_Year'] == 1983]
companies_highest_marketcap_1983 = companies_1983.loc[companies_1983['Market_Cap'].idxmax()]['Name']
companies_highest_marketcap_1983


'Amgen Inc.'

... Amgen Inc. has the highest Market Capitalization.

41. Set the column __Symbol__ as new Index of the __listings__ DataFrame! __Save__ the change!

In [9]:
listings.set_index('Symbol')

Exchange                          Name  Last_Price    Market_Cap  \
Symbol                                                                    
AMAT     NASDAQ       Applied Materials, Inc.       42.17  4.003589e+10   
COKE     NASDAQ  Coca-Cola Consolidated, Inc.      294.70  2.104584e+09   
WDFC     NASDAQ                 WD-40 Company      163.60  2.257237e+09   
AAPL     NASDAQ                    Apple Inc.      200.62  9.459795e+11   
KLAC     NASDAQ        KLA-Tencor Corporation      123.34  2.018632e+10   
...         ...                           ...         ...           ...   
ZB^A       NYSE     Zions Bancorporation N.A.       23.70           NaN   
ZB^G       NYSE     Zions Bancorporation N.A.       27.62           NaN   
ZB^H       NYSE     Zions Bancorporation N.A.       25.55           NaN   
ZIONW    NASDAQ     Zions Bancorporation N.A.       14.74           NaN   
ZTEST    NASDAQ             BATS BZX Exchange       10.10           NaN   

        ADR TSO  IPO_Year                 Sector  \
Symbol                                             
AMAT        NaN    1972.0             Technology   
COKE        NaN    1972.0  Consumer Non-Durables   
WDFC        NaN    1973.0       Basic Industries   
AAPL        NaN    1980.0             Technology   
KLAC        NaN    1980.0          Capital Goods   
...         ...       ...                    ...   
ZB^A        NaN       NaN                    NaN   
ZB^G        NaN       NaN                    NaN   
ZB^H        NaN       NaN                    NaN   
ZIONW       NaN       NaN                Finance   
ZTEST       NaN       NaN                    NaN   

                                   Industry  
Symbol                                       
AMAT                         Semiconductors  
COKE    Beverages (Production/Distribution)  
WDFC                        Major Chemicals  
AAPL                 Computer Manufacturing  
KLAC                  Electronic Components  
...                                     ...  
ZB^A                                    NaN  
ZB^G                                    NaN  
ZB^H                                    NaN  
ZIONW                           Major Banks  
ZTEST                                   NaN  

[6852 rows x 8 columns]

42. __Sort__ the __listings__ DataFrame by the new __index__ from __low to high__. __Save__ the new order!

In [10]:
listings.sort_index(ascending= True,inplace= True)

In [11]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [12]:
# run the cell! (create RangeIndex)
listings.reset_index(inplace = True)

In [13]:
# run the cell!
listings.head()

index Symbol Exchange                           Name  Last_Price  \
0      0      A     NYSE     Agilent Technologies, Inc.       81.68   
1      1     AA     NYSE              Alcoa Corporation       29.15   
2      2   AABA   NASDAQ                    Altaba Inc.       75.39   
3      3    AAC     NYSE             AAC Holdings, Inc.        2.16   
4      4    AAL   NASDAQ  American Airlines Group, Inc.       34.02   

     Market_Cap  ADR TSO  IPO_Year            Sector  \
0  2.593470e+10      NaN    1999.0     Capital Goods   
1  5.407810e+09      NaN    2016.0  Basic Industries   
2  4.278113e+10      NaN       NaN        Technology   
3  5.314109e+07      NaN    2014.0       Health Care   
4  1.527687e+10      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

43. Check the number of __unique values__ in __all columns__ of the __listings__ DataFrame! __How many__ unique values are in the __"Exchange"__ column?

In [14]:
listings.nunique()


index         6852
Symbol        6840
Exchange         3
Name          5766
Last_Price    4470
Market_Cap    5835
ADR TSO        140
IPO_Year        42
Sector          12
Industry       135
dtype: int64

There are ... 3 unique values in the "Exchange" column!

44. __Select__ the __10__ listings with the __highest Market_Cap__! What is the Company __Name__ of the __third largest__ listing?

In [38]:
top_3_highest_marketcap = listings.nlargest(columns='Market_Cap',n = 3)
third_largest_company = top_3_highest_marketcap.loc[top_3_highest_marketcap['Market_Cap'].idxmin()]['Name']
third_largest_company


'Amazon.com, Inc.'

The third largest company in terms of Market Capitalization is ... Amazon!

45. __Select__ the __15__ listings with the __lowest Prices__ (Penny Stocks)! Which Company has the __fourth lowest__ Stock Price?

In [23]:
lowest_prices_rows= listings.nsmallest(n=4,columns = "Last_Price")
res = lowest_prices_rows.loc[lowest_prices_rows['Market_Cap'].idxmin()]['Name']
res
# res = lowest_prices_rows[lowest_prices_rows['Market_Cap'].idxmin()]['Name']
# res

KeyError: nan

The fourth lowest Stock Price has ... Wheeler Real Estate Investment Trust!

In [24]:
# run the cell!
listings.head()

index Symbol Exchange                           Name  Last_Price  \
0      0      A     NYSE     Agilent Technologies, Inc.       81.68   
1      1     AA     NYSE              Alcoa Corporation       29.15   
2      2   AABA   NASDAQ                    Altaba Inc.       75.39   
3      3    AAC     NYSE             AAC Holdings, Inc.        2.16   
4      4    AAL   NASDAQ  American Airlines Group, Inc.       34.02   

     Market_Cap  ADR TSO  IPO_Year            Sector  \
0  2.593470e+10      NaN    1999.0     Capital Goods   
1  5.407810e+09      NaN    2016.0  Basic Industries   
2  4.278113e+10      NaN       NaN        Technology   
3  5.314109e+07      NaN    2014.0       Health Care   
4  1.527687e+10      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [25]:
#run the cell!
listings.tail()

index Symbol Exchange                           Name  Last_Price  \
6847   6847   ZUMZ   NASDAQ                    Zumiez Inc.       26.72   
6848   6848    ZUO     NYSE                    Zuora, Inc.       19.79   
6849   6849   ZYME     NYSE                 Zymeworks Inc.       15.74   
6850   6850   ZYNE   NASDAQ  Zynerba Pharmaceuticals, Inc.        7.85   
6851   6851   ZYXI   NASDAQ                    Zynex, Inc.        5.02   

        Market_Cap  ADR TSO  IPO_Year             Sector  \
6847  6.819191e+08      NaN    2005.0  Consumer Services   
6848  2.147266e+09      NaN    2018.0         Technology   
6849  5.040782e+08      NaN    2017.0        Health Care   
6850  1.653995e+08      NaN    2015.0        Health Care   
6851  1.618349e+08      NaN       NaN        Health Care   

                                               Industry  
6847                     Clothing/Shoe/Accessory Stores  
6848            Computer Software: Prepackaged Software  
6849                              Major Pharmaceuticals  
6850                              Major Pharmaceuticals  
6851  Biotechnology: Electromedical & Electrotherape...

46. __Check__ for all elements in the column __"Exchange"__, whether they are __equal to "NASDAQ"__. __Save__ the boolean Series in the variable __mask1__! <br>
__Fill in the gaps__!

In [26]:
mask1 = listings.Exchange == "NASDAQ"
mask1

0       False
1       False
2        True
3       False
4        True
        ...  
6847     True
6848    False
6849    False
6850     True
6851     True
Name: Exchange, Length: 6852, dtype: bool

47. __Check__ for all elements in the column __"Last_Price"__, whether they are __less than 1.0__. __Save__ the boolean Series in the variable __mask2__!

In [27]:
mask2 = listings.Last_Price < 1.0
mask2

0       False
1       False
2       False
3       False
4       False
        ...  
6847    False
6848    False
6849    False
6850    False
6851    False
Name: Last_Price, Length: 6852, dtype: bool

48. __Filter__ the DataFrame __listings__ for all __NASDAQ-listings__ (mask1) and __save__ the resulting DataFrame in the variable __nasdaq__ (create a copy)! <br>
__Fill in the gaps__!

In [28]:
nasdaq = listings.loc[mask1].copy()

__How many NASDAQ-listings__ does the __nasdaq__ DataFrame contain?

In [29]:
# run the cell!
nasdaq.head()

index Symbol Exchange                           Name  Last_Price  \
2      2   AABA   NASDAQ                    Altaba Inc.       75.39   
4      4    AAL   NASDAQ  American Airlines Group, Inc.       34.02   
6      6   AAME   NASDAQ  Atlantic American Corporation        2.48   
8      8   AAOI   NASDAQ  Applied Optoelectronics, Inc.       12.30   
9      9   AAON   NASDAQ                     AAON, Inc.       44.88   

     Market_Cap  ADR TSO  IPO_Year          Sector  \
2  4.278113e+10      NaN       NaN      Technology   
4  1.527687e+10      NaN       NaN  Transportation   
6  4.998398e+07      NaN       NaN         Finance   
8  2.445561e+08      NaN    2013.0      Technology   
9  2.336955e+09      NaN       NaN   Capital Goods   

                          Industry  
2                     EDP Services  
4    Air Freight/Delivery Services  
6                   Life Insurance  
8                   Semiconductors  
9  Industrial Machinery/Components

In [30]:
# run the cell!
nasdaq.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3435 entries, 2 to 6851
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   index       3435 non-null   int64  
 1   Symbol      3435 non-null   object 
 2   Exchange    3435 non-null   object 
 3   Name        3435 non-null   object 
 4   Last_Price  3432 non-null   float64
 5   Market_Cap  3168 non-null   float64
 6   ADR TSO     140 non-null    float64
 7   IPO_Year    1599 non-null   float64
 8   Sector      2964 non-null   object 
 9   Industry    2964 non-null   object 
dtypes: float64(4), int64(1), object(5)
memory usage: 295.2+ KB


There are in total ... 3435 NASDAQ-listings!

49. __Filter__ the DataFrame __listings__ for all __Penny Stocks__ (mask2) and __save__ the resulting DataFrame in the variable __penny__ (create a copy)!

In [31]:
penny = listings.loc[mask2].copy()

__How many Penny Stocks__ does the __penny__ DataFrame contain?

In [32]:
# run the cell!
penny.head()

index Symbol Exchange                        Name  Last_Price  \
13      13    AAU     AMEX      Almaden Minerals, Ltd.      0.5750   
67      67   ACST   NASDAQ         Acasti Pharma, Inc.      0.9500   
81      81  ADILW   NASDAQ  Adial Pharmaceuticals, Inc      0.3705   
87      87   ADOM   NASDAQ               ADOMANI, Inc.      0.3426   
115    115   AEMD   NASDAQ       Aethlon Medical, Inc.      0.8300   

       Market_Cap  ADR TSO  IPO_Year            Sector  \
13   6.424286e+07      NaN    2015.0  Basic Industries   
67   7.422610e+07      NaN       NaN       Health Care   
81            NaN      NaN    2018.0       Health Care   
87   2.495277e+07      NaN    2017.0     Capital Goods   
115  1.573722e+07      NaN       NaN     Capital Goods   

                                             Industry  
13                                    Precious Metals  
67                              Major Pharmaceuticals  
81                              Major Pharmaceuticals  
87                                  Auto Parts:O.E.M.  
115  Biotechnology: Laboratory Analytical Instruments

In [33]:
# run the cell!
penny.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 327 entries, 13 to 6840
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   index       327 non-null    int64  
 1   Symbol      327 non-null    object 
 2   Exchange    327 non-null    object 
 3   Name        327 non-null    object 
 4   Last_Price  327 non-null    float64
 5   Market_Cap  199 non-null    float64
 6   ADR TSO     6 non-null      float64
 7   IPO_Year    161 non-null    float64
 8   Sector      324 non-null    object 
 9   Industry    324 non-null    object 
dtypes: float64(4), int64(1), object(5)
memory usage: 28.1+ KB


There are in total ... 327 Penny Stocks!

50. __Filter__ the DataFrame __listings__ for all __Penny Stocks__ (mask2) that are __NASDAQ-listed__ (mask1) and __save__ the resulting DataFrame in the variable __nasdaq_penny__! __Fill in the gaps!__

In [34]:
nasdaq_penny = listings.loc[mask1 & mask2]

__How many listings__ does the __nasdaq_penny__ DataFrame contain?

In [35]:
# run the cell!
nasdaq_penny.head()

index Symbol Exchange                                  Name  Last_Price  \
67      67   ACST   NASDAQ                   Acasti Pharma, Inc.      0.9500   
81      81  ADILW   NASDAQ            Adial Pharmaceuticals, Inc      0.3705   
87      87   ADOM   NASDAQ                         ADOMANI, Inc.      0.3426   
115    115   AEMD   NASDAQ                 Aethlon Medical, Inc.      0.8300   
122    122   AETI   NASDAQ  American Electric Technologies, Inc.      0.8741   

       Market_Cap  ADR TSO  IPO_Year         Sector  \
67   7.422610e+07      NaN       NaN    Health Care   
81            NaN      NaN    2018.0    Health Care   
87   2.495277e+07      NaN    2017.0  Capital Goods   
115  1.573722e+07      NaN       NaN  Capital Goods   
122  7.927501e+06      NaN       NaN         Energy   

                                             Industry  
67                              Major Pharmaceuticals  
81                              Major Pharmaceuticals  
87                                  Auto Parts:O.E.M.  
115  Biotechnology: Laboratory Analytical Instruments  
122                   Industrial Machinery/Components

In [36]:
# run the cell!
nasdaq_penny.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 257 entries, 67 to 6836
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   index       257 non-null    int64  
 1   Symbol      257 non-null    object 
 2   Exchange    257 non-null    object 
 3   Name        257 non-null    object 
 4   Last_Price  257 non-null    float64
 5   Market_Cap  131 non-null    float64
 6   ADR TSO     6 non-null      float64
 7   IPO_Year    144 non-null    float64
 8   Sector      256 non-null    object 
 9   Industry    256 non-null    object 
dtypes: float64(4), int64(1), object(5)
memory usage: 22.1+ KB


There are in total ... 257 penny stocks at nasdaq!

# Well Done!

------------------------------

# Hints (Spoiler!)

39. sort_values() method, by parameter

40. pass lists to the by and ascending parameter

41. set_index() method, inplace

42. sort_index() method, inplace

43. nunique() method

44. nlargest() method, parameters n and columns

45. nsmallest() method

46. listings.--- == "---"

47. listings.--- < ---

48. pass mask1 to the loc operator

49. listings.loc[---].---

50. listings.loc[--- & ---]